private/bufpkg/bufcas/manifest.go

```diff
@@ -0,0 +1,183 @@
+		}
+		fileNodes = append(fileNodes, fileNode)
+	}
+	return NewManifest(fileNodes)
+}
+
+// ManifestToBlob converts the string representation of the given Manifest into a Blob.
+//
+// The Manifest is assumed to be non-nil.
+func ManifestToBlob(manifest Manifest) (Blob, error) {
+	return NewBlobForContent(DigestTypeShake256, strings.NewReader(manifest.String()))
+}
+
+// BlobToManifest converts the given Blob representing the string representation of a Manifest into a Manifest.
+//
+// The Blob is assumed to be non-nil
+func BlobToManifest(blob Blob) (Manifest, error) {
+	return NewManifestForString(string(blob.Content()))
+}
+
+// ManifestToProtoBlob converts the string representation of the given Manifest into a proto Blob.
+//
+// The Manifest is assumed to be non-nil.
+//
+// TODO: validate the returned proto Blob.
+func ManifestToProtoBlob(manifest Manifest) (*storagev1beta1.Blob, error) {
+	blob, err := ManifestToBlob(manifest)
+	if err != nil {
+		return nil, err
+	}
+	return BlobToProto(blob)
+}
+
+// BlobToManifest converts the given proto Blob representing the string representation of a Manifest into a Manifest.
+//
+// # The proto Blob is assumed to be non-nil
+//
+// TODO: validate the input proto Blob.
+func ProtoBlobToManifest(protoBlob *storagev1beta1.Blob) (Manifest, error) {
+	blob, err := ProtoToBlob(protoBlob)
+	if err != nil {
+		return nil, err
+	}
+	return BlobToManifest(blob)
+}
+
+// *** PRIVATE ***
+
+type manifest struct {
+	pathToFileNode        map[string]FileNode
+	sortedUniqueFileNodes []FileNode
+}
+
+func newManifest(fileNodes []FileNode) (*manifest, error) {
+	pathToFileNode := make(map[string]FileNode)
+	for _, fileNode := range fileNodes {
+		if existingFileNode, ok := pathToFileNode[fileNode.Path()]; ok {
+			if !DigestEqual(existingFileNode.Digest(), fileNode.Digest()) {
+				return nil, fmt.Errorf("path %q had different Digests when constructing FileNode", fileNode.Path())
+			}
+		} else {
+			pathToFileNode[fileNode.Path()] = fileNode
+		}
+	}
+	// Just cache ahead of time for now.
+	sortedUniqueFileNodes := make([]FileNode, 0, len(pathToFileNode))
+	for _, fileNode := range pathToFileNode {
+		sortedUniqueFileNodes = append(sortedUniqueFileNodes, fileNode)
+	}
+	sort.Slice(
+		sortedUniqueFileNodes,
+		func(i int, j int) bool {
+			return sortedUniqueFileNodes[i].Path() < sortedUniqueFileNodes[j].Path()
+		},
+	)
+	return &manifest{
+		pathToFileNode:        pathToFileNode,
+		sortedUniqueFileNodes: sortedUniqueFileNodes,
+	}, nil
+}
+
+func (m *manifest) FileNodes() []FileNode {
+	return m.sortedUniqueFileNodes
+}
+
+func (m *manifest) GetDigest(path string) Digest {
+	fileNode, ok := m.pathToFileNode[path]
+	if !ok {
+		return nil
+	}
+	return fileNode.Digest()
+}
+
+func (m *manifest) String() string {
+	buffer := bytes.NewBuffer(nil)
+	for _, fileNode := range m.sortedUniqueFileNodes {
+		_, _ = buffer.WriteString(fileNode.String())
+		_, _ = buffer.WriteRune('\n')
+	}
+	return buffer.String()
```

Comment:
nit: Since we're going to use this a lot, we might switch to `strings.Builder` instead. It is documented to do less copying that bytes.Buffer for building strings.

Ref: https://pkg.go.dev/bytes#Buffer.String